Integrate Debugger within Collab DB

Anthony-Eid opened the pull request:
This PR integrates Zed's Debugger with the Collab database, enabling Zed to guarantee that clients joining a project with active debug sessions in progress will receive and set up those sessions.

Initially, I planned to implement a GetDebuggerSession RPC that would enable clients joining a session in the middle of debugging to get the active debugger state without involving the Collab database. However, Mikayla advised me to treat the Collab database as RAM, prompting me to alter my approach to achieving this.

Tasks Completed:
- Create debug_clients table in Collab db
- Create debug_panel_items table in Collab db
- Send debugger session with a join_project_response
- Implement a unit test
- Sync Collab DB with debugger state by implementing the below functionality
   - Handle adapter shutdown
   - Handle DebugAdapterUpdate RPC
   - Handle SetDebuggerPanelItem RPC
   - Handle SetDebuggerCapabilities RPC
 

### Demo
https://github.com/user-attachments/assets/dd3e1160-73a3-4ae0-9f2b-840990890126

Anthony-Eid committed:
Add test for setting active debug panel items on project join

Anthony-Eid committed:
Add DebuggerSession proto message

Anthony-Eid committed:
Modify debugger session

Anthony-Eid committed:
Merge with debugger

Anthony-Eid committed:
Get collab server to build

Anthony-Eid committed:
Get collab test to compile

Anthony-Eid committed:
Add proto messages

Anthony-Eid committed:
Set up message handler for get debugger sessions

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -1569,6 +1576,20 @@ impl DapStore {
+        mut cx: AsyncAppContext,
+    ) -> Result<()> {
+        this.update(&mut cx, |dap_store, cx| {
+            if let Some((_, _)) = dap_store.downstream_client {
```
Why do we need to check here if we have a downstream client?

Anthony-Eid replied:
Just a sanity check. I don’t think it will ever fail but just in guess theirs a future regression or something we don’t think about.

RemcoSmitsDev replied:
Uhm but this code is never hit inside the if statement right? Because we should only end up inside this method on the non host clients, so there will never be a downstream client. Or am I missing something?

Anthony-Eid replied:
Shoot you’re right!! I should’ve use upstream_client instead. I was up late last night and didn’t realize.

RemcoSmitsDev replied:
All good, I believe this is already covered by the collab code that it's only called when you are actually inside a shared collab channel.

Anthony-Eid replied:
The check was supposed to ensure that only the host would respond to a GetDebuggerSessions message. I wasn't sure if a message is propagated to all clients or just the host.

Anthony-Eid replied:
I tested this and the original downstream_client was correct. We only want to send active sessions if we have downstream clients

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -1569,6 +1576,20 @@ impl DapStore {
+        this.update(&mut cx, |dap_store, cx| {
+            if let Some((_, _)) = dap_store.downstream_client {
+                cx.emit(DapStoreEvent::SendDebuggerSessions(ProjectId(
+                    _envelope.payload.project_id,
```
What does this event `SendDebuggerSessions` do? No one is acting on it.

Anthony-Eid replied:
Debug panel is supposed to act on it to send active sessions. Im planning on adding a client handler directly to the debug panel to clean this up

RemcoSmitsDev commented on crates/project/src/dap_store.rs:
```diff
@@ -170,6 +173,10 @@ impl DapStore {
         upstream_client: AnyProtoClient,
         _: &mut ModelContext<Self>,
     ) -> Self {
+        upstream_client
```
Can we move this to the `shared` method? This feels a bit hacky solution. So we also don't try to send a proto message when you are not event sharing a project.

Anthony-Eid replied:
Yes

RemcoSmitsDev commented on crates/collab/src/tests/debug_panel_tests.rs:
```diff
@@ -163,3 +163,136 @@ async fn test_debug_panel_item_opens_on_remote(
+
+#[gpui::test]
+async fn test_active_debug_panel_item_set_on_join_project(
+    cx_a: &mut TestAppContext,
```
Great to see more tests😀

Anthony-Eid replied:
At least one test per new feature!

Anthony-Eid committed:
Send set debug panel requests when handling get debugger sessions msg

Anthony-Eid committed:
Get request to send and set debug panel

Anthony-Eid committed:
Merge branch 'debugger' into join-project-during-debug-session

Anthony-Eid committed:
Setup ground work for debug sessions collab db table

Anthony-Eid committed:
Set up debug_client table for collab db

Anthony-Eid committed:
Remove up GetDebuggerSession proto request code

Anthony-Eid committed:
Get proto::DebuggerSession during join_project_internal

Anthony-Eid committed:
Remove debug_sessions table from collab db

Anthony-Eid committed:
Add migration for debug_client and fix some bugs

RemcoSmitsDev commented on crates/collab/migrations/20250107082721_create_debug_client_table.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE debug_clients (
```
Don't forget to add this to the test sql file!

Anthony-Eid replied:
For sure! I'm planning on changing it more so I'll update the test file when it's stable

Anthony-Eid committed:
Create dap store event queue for remote daps

When creating a project in from_join_project_response(...) the debug panel hasn't been initialized yet so
it can't handle dap_store events. The solution to this is creating an event queue that debug panel takes
from dap store if it's remote and then handles all the events

Anthony-Eid committed:
Fix debug panel join project during session test

Anthony-Eid committed:
Add debug_panel_items table to collab db

Anthony-Eid committed:
Integrate debug_panel_item table into collab msg handlers

Anthony-Eid committed:
Finialize debug_panel_item table refactor for collab db

Anthony-Eid committed:
Integrate UpdateDebugAdapter RPC with collab DB

Anthony-Eid committed:
Handle ShutdownDebugClient RPC for collab db

Anthony-Eid committed:
Merge branch 'debugger' into join-project-during-debug-session

Anthony-Eid committed:
Fix clippy
